libfreerdp/core/gateway/arm.c

```diff
@@ -0,0 +1,348 @@
+
+	if (!arm || !method)
+		return NULL;
+
+	uri = http_context_get_uri(arm->http);
+	request = http_request_new();
+
+	if (!request)
+		return NULL;
+
+	if (!http_request_set_method(request, method) || !http_request_set_uri(request, uri))
+		goto out;
+
+	else if (arm->settings->GatewayHttpExtAuthBearer)
+	{
+		http_request_set_auth_scheme(request, "Bearer");
+		http_request_set_auth_param(request, arm->settings->GatewayHttpExtAuthBearer);
+	}
+
+	http_request_set_transfer_encoding(request, transferEncoding);
+	http_request_set_content_length(request, content_length);
+	http_request_set_content_type(request, content_type);
+
+	s = http_request_write(arm->http, request);
+out:
+	http_request_free(request);
+
+	if (s)
+		Stream_SealLength(s);
+
+	return s;
+}
+
+static BOOL arm_send_http_request(rdpArm* arm, rdpTls* tls, const char* method,
+                                  const char* content_type, const char* data, size_t content_length)
+{
+	size_t sz;
+	wStream* s = NULL;
+	int status = -1;
+	s = arm_build_http_request(arm, method, TransferEncodingIdentity, content_type, content_length);
+
+	if (!s)
+		return FALSE;
+
+	sz = Stream_Length(s);
+
+	if (sz <= INT_MAX)
+		status = freerdp_tls_write_all(tls, Stream_Buffer(s), (int)sz);
+
+	Stream_Free(s, TRUE);
+	if (status >= 0 && content_length > 0)
+	{
+		status = freerdp_tls_write_all(tls, (const BYTE*)data, content_length);
+	}
+	return (status >= 0);
+}
+
+static void arm_free(rdpArm* arm)
+{
+	if (!arm)
+		return;
+
+	freerdp_tls_free(arm->tls);
+	http_context_free(arm->http);
+
+	free(arm);
+}
+
+BOOL arm_resolve_endpoint(rdpContext* context, DWORD timeout)
+{
+#ifndef WITH_CJSON
+	WLog_ERR(TAG, "arm gateway support not compiled in");
+	return FALSE;
+#else
+	if (!context)
+		return FALSE;
+
+	if ((context->settings->LoadBalanceInfoLength == 0) ||
+	    (context->settings->RemoteApplicationProgram == NULL))
+		return FALSE;
+
+	rdpArm* arm = (rdpArm*)calloc(1, sizeof(rdpArm));
+	char* lbi = NULL;
+	cJSON* json = NULL;
+	char* message = NULL;
+
+	if (arm)
+	{
+		arm->context = context;
+		arm->settings = arm->context->settings;
+		arm->tls = freerdp_tls_new(arm->settings);
+		HttpResponse* response = NULL;
+		long StatusCode;
+
+		if (!arm->tls)
+			goto arm_error;
+
+		arm->http = http_context_new();
+
+		if (!arm->http)
+			goto arm_error;
+
+		if (!http_context_set_uri(arm->http, "/api/arm/v2/connections/") ||
+		    !http_context_set_accept(arm->http, "application/json") ||
+		    !http_context_set_cache_control(arm->http, "no-cache") ||
+		    !http_context_set_pragma(arm->http, "no-cache") ||
+		    !http_context_set_connection(arm->http, "Keep-Alive") ||
+		    !http_context_set_user_agent(arm->http, "FreeRDP/3.0") ||
+		    !http_context_set_x_ms_user_agent(arm->http, "FreeRDP/3.0") ||
+		    !http_context_set_host(arm->http, arm->settings->GatewayHostname))
+			goto arm_error;
+
+		if (!arm_tls_connect(arm, arm->tls, timeout))
+			goto arm_error;
+
+		json = cJSON_CreateObject();
```

Comment:
will not compile without `cJSON`, better move the guards so everything is covered.

Reply:
actually the "whole" function body is `#ifdef CJSON`
else it will print
`WLog_ERR(TAG, "arm gateway support not compiled in");`
and
immediatly return FALSE

see line 220

Reply:
oh, indeed, `#else`.
still, might be a good idea to just not compile in the whole code for this if `cJSON` is not available, not just fail here. (but can be added in a separate pr, this one is already huge enough)